vowpalwabbit/gd.cc

```diff
@@ -495,8 +495,6 @@ inline void pred_per_update_feature(norm_data& nd, float x, float& fw)
       x = (x > 0) ? x_min : -x_min;
       x2 = x2_min;
     }
-    if (x2 > x2_max)
-      THROW("your features have too much magnitude");
```

Comment:
don't we have decent chance to turn w[adaptive] to inf on line 506 now?

Reply:
I suspect 'yes', but I'm not seeing how that is a problem.  Do you?

Reply:
Once it is there, it stays forever => we are not going to learn on next occurences of this feature forever. Or am I missing something?

I tried that test:

Python code to generate data:
n = 1000
passes = 10

#healthy data that mimicking y=sign(x1^4 - x2)
def write_good_data(f, n, passes):
    for i in range(n):
        for k in range(passes):
            label = -1 if i % 2 == 0 else 1
            x = i / n
            delta = (x**4) * 0.1 * label
            f.write(f'{label} |a b:{x} |c b:{x**4 + delta}'+'\n')


good_path = rf'good-{n}.txt'
bad_path = rf'bad-{n}.txt'

#bad data: one poisoned example + long tail of healthy y = sign(x1^4-x2) data
with open(bad_path, 'w') as f:
    f.write('1 |a b:10000000000')
    write_good_data(f, n, passes)

#good data: only long tail of healthy y = sign(x1^4-x2) data
with open(good_path, 'w') as f:
    write_good_data(f, n, passes)


Execution:
.\vw --interactions aaaa --noconstant -d bad-100.txt --invert_hash readable_bad.txt
.\vw --interactions aaaa --noconstant -d good-100.txt --invert_hash readable_good.txt

Result:
readable_good.txt:
a^b*a^b*a^b*a^b:68928:-1.10958
a^b:144016:-0.67623
c^b:196862:2.12328

readable_bad.txt:
:0
a^b:144016:-2.17053e-19
c^b:196862:0.322327

Reply:
This seems like an appropriate failure mode.